libr/anal/xrefs.c

```diff
@@ -537,20 +283,84 @@ R_API const char *r_anal_xrefs_type_tostring (char type) {
 	}
 }
 
-typedef struct {
-	RAnal *anal;
-	int count;
-} CountState;
+R_API bool r_anal_xrefs_init(RAnal *anal) {
+	SdbHash *tmp;
 
-static int countcb(CountState *cs, const char *k, const char *v) {
-	if (!strncmp (k, "ref.", 4)) {
-		cs->count ++;
+	ht_free (anal->dict_refs);
+	anal->dict_refs = NULL;
+	ht_free (anal->dict_xrefs);
+	anal->dict_xrefs = NULL;
+
+	tmp = ht_new (NULL, xrefs_ht_free, NULL);
+	if (!tmp) {
+		return false;
 	}
-	return 1;
+	anal->dict_refs = tmp;
+
+	tmp = ht_new (NULL, xrefs_ht_free, NULL);
+	if (!tmp) {
+		ht_free (anal->dict_refs);
+		anal->dict_refs = NULL;
+		return false;
+	}
+	anal->dict_xrefs = tmp;
+	return true;
 }
 
 R_API int r_anal_xrefs_count(RAnal *anal) {
-	CountState cs = { anal, 0 };
-	sdb_foreach (DB, (SdbForeachCallback)countcb, &cs);
-	return cs.count;
+	return anal->dict_xrefs->count;
+}
+
+static int ref_cmp(const RAnalRef *a, const RAnalRef *b) {
+	if (a->at < b->at) {
+		return -1;
+	} else if (a->at > b->at) {
+		return 1;
+	} else {
+		if (a->addr < b->addr) {
+			return -1;
+		} else if (a->addr > b->addr) {
+			return 1;
+		}
+	}
+	return 0;
+}
+
+static RList *fcn_get_refs(RAnalFunction *fcn, SdbHash *ht) {
+	RListIter *iter;
+	RAnalBlock *bb;
+	RList *list = r_anal_ref_list_new ();
+	if (!list) {
+		return NULL;
+	}
+
+	r_list_foreach (fcn->bbs, iter, bb) {
+		int i;
+
+		for (i = 0; i < bb->ninstr; ++i) {
+			ut64 at = bb->addr + r_anal_bb_offset_inst (bb, i);
+			listxrefs (ht, at, list);
+		}
+	}
+	return list;
+}
+
+R_API RList *r_anal_fcn_get_refs(RAnal *anal, RAnalFunction *fcn) {
+	return fcn_get_refs (fcn, anal->dict_refs);
+}
+
+R_API RList *r_anal_fcn_get_xrefs(RAnal *anal, RAnalFunction *fcn) {
+	return fcn_get_refs (fcn, anal->dict_xrefs);
+}
+
+R_API RList *r_anal_fcn_get_refs_sorted(RAnal *anal, RAnalFunction *fcn) {
+	RList *l = r_anal_fcn_get_refs (anal, fcn);
```

Comment:
this is a oneliner, dont waste 3 lines, imho return r_list_sort (r_anal_fcn_get_Refs (anal, fcn), ref_cmp) should be good. 

also the cast for the function signature shouldnt be needed if the function gets void*

Reply:
1) r_list_sort unfortunately doesn't return a RList*, so you can't really do it in one line
2) the cast for the function signature is needed because i want to directly use RAnalRef as parameters. If you do otherwise, the signature should be (void*,void*) and you need to cast *both* arguments inside the function, which seems useless when you can just cast the function signature, as I did.

Reply:
oops right about the rlistSort() meh its fine yep i know about this caveat, but casting a function signature is more dangerous than casting the arguments because you can pass a wrong callback and didnt notice until running., if the funciton have different numnber of arguments or types it will complain at compile time. i can merge like this , it was just a comment :)